src/wiseagents/wise_agent_impl.py

```diff
@@ -632,6 +636,316 @@ def set_response_delivery(self, response_delivery: Callable[[], WiseAgentMessage
+        if len(ctx.get_required_agents_for_current_phase(chat_id)) == 0:
+            next_phase = ctx.get_next_phase(chat_id)
+            if next_phase is None:
+                # Determine the final answer
+                final_answer_prompt = ("What is the final answer for the original query? Provide the answer followed" +
+                                       " by a confidence score from 0 to 100 to indicate how certain you are of the" +
+                                       " answer. Format the response with just the answer first followed by just" +
+                                       " the confidence score on the next line. For example:\n" +
+                                       " Your answer goes here.\n"
+                                       " 85\n")
+                ctx.append_chat_completion(chat_uuid=chat_id,
+                                           messages={"role": "user", "content": final_answer_prompt})
+                llm_response = self.llm.process_chat_completion(ctx.llm_chat_completion[chat_id], tools=[])
+                final_answer_and_score = llm_response.choices[0].message.content.splitlines()
+                final_answer = "\n".join(final_answer_and_score[:-1])
+                if final_answer_and_score[-1].strip().isnumeric():
+                    score = int(final_answer_and_score[-1])
+                else:
+                    # A score could not be determined
+                    score = 0
+
+                # Determine if we should return the final answer or iterate
+                if score >= self.confidence_score_threshold:
+                    self.send_response(WiseAgentMessage(message=final_answer, sender=self.name,
+                                                        context_name=response.context_name, chat_id=chat_id), self._route_response_to)
+                elif len(ctx.get_queries(chat_id)) == self.max_iterations:
+                    self.send_response(WiseAgentMessage(message=CANNOT_ANSWER, sender=self.name,
+                                                        context_name=response.context_name, chat_id=chat_id),
+                                       self._route_response_to)
+                else:
+                    # Rephrase the query and iterate
+                    if len(ctx.get_queries(chat_id)) < self.max_iterations:
+                        rephrase_query_prompt = ("The final answer was not considered good enough to respond to the original query.\n" +
+                                                 " The original query was: " + ctx.get_queries(chat_id)[0] + "\n" +
+                                                 " Your task is to analyze the original query for its intent and rephrase it in a way"
+                                                 " to yield a better final answer. The response should contain only the rephrased query."
+                                                 " Don't include anything else in the response.\n")
```

Comment:
So here we are loosing all the previous answers? Or do we keep in the llm messages and may be used by other agents?

A more sophisticated (to be tracked with an issue or to have another github discussion), is to understand why we don't get the "fggod enough" answer and see if there is part of the previous agents' answers to be kept and part to be discarted. It may drive also to a phase where not all agents are reiterated, but only the ones which didn't contributed enough to a "good answer".
Not certeninly blocking this PR...just food for thoughts

Reply:
Currently, we don't lose the previous answers, they're still part of the llm_chat_completions so now the LLM will receive the previous messages along with the rephrased query. But this approach isn't ideal since the length can get quite long. https://github.com/wise-agents/wise-agents/issues/204 tracks the ability to summarized the share context but we do also need an issue to track this part (i.e., better determination of which agents to execute again with a rephrased query and what parts of the shared context to discard).

Reply:
Created https://github.com/wise-agents/wise-agents/issues/215 to track this.